viewing_party/party.py

```diff
@@ -0,0 +1,255 @@
+
+def add_to_watchlist(user_data,movie):
+    user_data["watchlist"].append(movie)
+
+    return user_data  
+
+def watch_movie(user_data, title): 
+    watch_list = user_data["watchlist"] 
+    watched = user_data["watched"]
+
+    #if title: #in movie["title"] :
+    for movie in watch_list:
+        if title in movie["title"]:
+            watch_list.remove(movie)
+            watched.append(movie)
+        
+    
+    return user_data
+    
+#Wave 2
+def get_watched_avg_rating(user_data):
+    total_rating = 0
+    if len(user_data["watched"]) == 0:
+        return 0
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"] 
+    return total_rating / len(user_data["watched"])
+
+
+
+def get_most_watched_genre(user_data):
+    genre_to_count_dict = {}
+#dict = {genre : count} 
+    if len(user_data["watched"]) == 0:
+        return None
+    for movie in user_data["watched"]:
+        if movie["genre"] not in genre_to_count_dict:
+            genre_to_count_dict[movie["genre"]] = 1
+        else: 
+            genre_to_count_dict[movie["genre"]] += 1
+    
+    max_count = 0
+    most_popular_genre = ""
+    for genre, count in genre_to_count_dict.items():
+        if count > max_count:
+            max_count = count
+            most_popular_genre = genre
+    return most_popular_genre
+
+#Wave 3
+
+def get_unique_watched(user_data):
+    if len(user_data["watched"]) == 0:
+        return []
+    
+    user_movies = set()
+    for watched in user_data["watched"]:
+       # for title, movie_title in watched.items():
+            user_movies.add(watched["title"])
+    
+    friends_movies = set()
+    for friends_data in user_data["friends"]:
+        for watched in friends_data["watched"]:
+            #for title, movie_titles in watched.items():
+                friends_movies.add(watched["title"])
+
+    user_unique = user_movies - friends_movies
+    #for  watched in user_data["watched"]
+    #if title in user_data["watched"]:
+    # user_unique.append()
+    uni_user_movies = []
+    for  watched in user_data["watched"]:
+        if watched["title"] in user_unique:
+            uni_user_movies.append(watched)
+
+    return uni_user_movies
+
+def get_friends_unique_watched(user_data):
+    user_movies = set()
+    for watched in user_data["watched"]:
+        #for title, movie_title in watched.items():
+            user_movies.add(watched["title"])
+
+    
+    friends_movies = set()
+    for friends_data in user_data["friends"]:
+        for watched in friends_data["watched"]:
+            #for title, movie_titles in watched.items():
+                friends_movies.add(watched["title"])
+
+
+    friends_unique = friends_movies - user_movies
+
+    uni_friends_movies = []
+    for movie in friends_unique:
+        #if movie in friends_unique:
+        uni_friends_movies.append({"title": movie})
+        if len(uni_friends_movies) == 0:
+            return []
+    
+    return uni_friends_movies
+
```

Comment:
Because `uni_friends_movies` starts off as an empty list and grows based on `friends_unique`, you can rearrange your conditional so that you can return `uni_friends_movies` instead of a new empty list.

```suggestion
    uni_friends_movies = []
    if len(friends_unique) > 0:
        for movie in friends_unique:
            uni_friends_movies.append({"title": movie})
    
    return uni_friends_movies
```

Reply:
In order to reuse `get_friends_unique_watched()` in other waves, we need to consider how the friends watched dictionary changes. In wave 3, we assume that the friend's movies only have the `title` key, however, in wave 4, the movies have `title` and `host`. To make this function more flexible , we can iterate through the friends watched list again and check if the movie titles are not in `users_unique` set.

```suggestion

 if len(friends_unique) > 0:
        for friend in user_data["friends"]:  
            for movie in friend["watched"]:
                if movie["title"] not in user_movies and movie not in uni_friends_movies:
                    uni_friends_movies.append(movie)
        
    return uni_friends_movies

```